diff --git a/src/vs/platform/environment/node/environmentService.ts b/src/vs/platform/environment/node/environmentService.ts
index 15b5c20..3c562e6 100644
--- a/src/vs/platform/environment/node/environmentService.ts
+++ b/src/vs/platform/environment/node/environmentService.ts
@@ -67,7 +67,7 @@ function getCLIPath(execPath: string, appRoot: string, isBuilt: boolean): string
 
 	// macOS
 	if (isBuilt) {
-		return path.join(appRoot, 'bin', 'code');
+		return path.join(appRoot, 'bin', `${product.applicationName}`);
 	}
 
 	return path.join(appRoot, 'scripts', 'code-cli.sh');
