diff --git a/src/vs/workbench/browser/parts/editor/editor.ts b/src/vs/workbench/browser/parts/editor/editor.ts
index 6fd048f..5645190 100644
--- a/src/vs/workbench/browser/parts/editor/editor.ts
+++ b/src/vs/workbench/browser/parts/editor/editor.ts
@@ -35,6 +35,8 @@ export const DEFAULT_EDITOR_PART_OPTIONS: IEditorPartOptions = {
 	hasIcons: true, // 'vs-seti' is our default icon theme
 	enablePreview: true,
 	openPositioning: 'right',
+	openPositioningSortOrder: 'asc',
+	openPositioningSortRule: 'name-local',
 	openSideBySideDirection: 'right',
 	closeEmptyGroups: true,
 	labelFormat: 'default',
diff --git a/src/vs/workbench/browser/workbench.contribution.ts b/src/vs/workbench/browser/workbench.contribution.ts
index 3a017b9..8a4e689 100644
--- a/src/vs/workbench/browser/workbench.contribution.ts
+++ b/src/vs/workbench/browser/workbench.contribution.ts
@@ -167,9 +167,21 @@ const registry = Registry.as<IConfigurationRegistry>(ConfigurationExtensions.Con
 			},
 			'workbench.editor.openPositioning': {
 				'type': 'string',
-				'enum': ['left', 'right', 'first', 'last'],
+				'enum': ['left', 'right', 'first', 'last', 'sort'],
 				'default': 'right',
-				'markdownDescription': localize({ comment: ['This is the description for a setting. Values surrounded by single quotes are not to be translated.'], key: 'editorOpenPositioning' }, "Controls where editors open. Select `left` or `right` to open editors to the left or right of the currently active one. Select `first` or `last` to open editors independently from the currently active one.")
+				'markdownDescription': localize({ comment: ['This is the description for a setting. Values surrounded by single quotes are not to be translated.'], key: 'editorOpenPositioning' }, "Controls where editors open. Select `left` or `right` to open editors to the left or right of the currently active one. Select `first` or `last` to open editors independently from the currently active one. Select `sort` to open editors in alphabetical order.")
+			},
+			'workbench.editor.openPositioningSortOrder': {
+				'type': 'string',
+				'enum': ['asc', 'desc'],
+				'default': 'asc',
+				'description': localize('openPositioningSortOrder', "Controls in which direction the open editors are sorted.")
+			},
+			'workbench.editor.openPositioningSortRule': {
+				'type': 'string',
+				'enum': ['name-local', 'name-absolute', 'absolute'],
+				'default': 'name-local',
+				'description': localize('openPositioningSortRule', "Controls how the open editors are sorted.")
 			},
 			'workbench.editor.openSideBySideDirection': {
 				'type': 'string',
diff --git a/src/vs/workbench/common/editor.ts b/src/vs/workbench/common/editor.ts
index 8b7f255..70809b9 100644
--- a/src/vs/workbench/common/editor.ts
+++ b/src/vs/workbench/common/editor.ts
@@ -815,7 +815,9 @@ interface IEditorPartConfiguration {
 	enablePreviewFromQuickOpen?: boolean;
 	enablePreviewFromCodeNavigation?: boolean;
 	closeOnFileDelete?: boolean;
-	openPositioning?: 'left' | 'right' | 'first' | 'last';
+	openPositioning?: 'left' | 'right' | 'first' | 'last' | 'sort';
+	openPositioningSortOrder?: 'asc' | 'desc';
+	openPositioningSortRule?: 'name-local' | 'name-absolute' | 'absolute';
 	openSideBySideDirection?: 'right' | 'down';
 	closeEmptyGroups?: boolean;
 	revealIfOpen?: boolean;
diff --git a/src/vs/workbench/common/editor/editorGroupModel.ts b/src/vs/workbench/common/editor/editorGroupModel.ts
index 1bf2263..5d58108 100644
--- a/src/vs/workbench/common/editor/editorGroupModel.ts
+++ b/src/vs/workbench/common/editor/editorGroupModel.ts
@@ -4,7 +4,7 @@
  *--------------------------------------------------------------------------------------------*/
 
 import { Event, Emitter } from 'vs/base/common/event';
-import { IEditorInputFactoryRegistry, IEditorIdentifier, IEditorCloseEvent, GroupIdentifier, IEditorInput, EditorsOrder, EditorExtensions, IUntypedEditorInput } from 'vs/workbench/common/editor';
+import { IEditorInputFactoryRegistry, IEditorIdentifier, IEditorCloseEvent, GroupIdentifier, IEditorInput, EditorsOrder, EditorExtensions, IUntypedEditorInput, Verbosity } from 'vs/workbench/common/editor';
 import { EditorInput } from 'vs/workbench/common/editor/editorInput';
 import { SideBySideEditorInput } from 'vs/workbench/common/editor/sideBySideEditorInput';
 import { IInstantiationService } from 'vs/platform/instantiation/common/instantiation';
@@ -17,7 +17,8 @@ const EditorOpenPositioning = {
 	LEFT: 'left',
 	RIGHT: 'right',
 	FIRST: 'first',
-	LAST: 'last'
+	LAST: 'last',
+	SORT: 'sort'
 };
 
 export interface EditorCloseEvent extends IEditorCloseEvent {
@@ -108,7 +109,9 @@ export class EditorGroupModel extends Disposable {
 	private active: EditorInput | null = null;  // editor in active state
 	private sticky: number = -1; // index of first editor in sticky state
 
-	private editorOpenPositioning: ('left' | 'right' | 'first' | 'last') | undefined;
+	private editorOpenPositioning: ('left' | 'right' | 'first' | 'last' | 'sort') | undefined;
+	private editorOpenPositioningSortOrder: ('asc' | 'desc') | undefined;
+	private editorOpenPositioningSortRule: ('name-local' | 'name-absolute' | 'absolute') | undefined;
 	private focusRecentEditorAfterClose: boolean | undefined;
 
 	constructor(
@@ -134,6 +137,8 @@ export class EditorGroupModel extends Disposable {
 
 	private onConfigurationUpdated(): void {
 		this.editorOpenPositioning = this.configurationService.getValue('workbench.editor.openPositioning');
+		this.editorOpenPositioningSortOrder = this.configurationService.getValue('workbench.editor.openPositioningSortOrder');
+		this.editorOpenPositioningSortRule = this.configurationService.getValue('workbench.editor.openPositioningSortRule');
 		this.focusRecentEditorAfterClose = this.configurationService.getValue('workbench.editor.focusRecentEditorAfterClose');
 	}
 
@@ -212,6 +217,63 @@ export class EditorGroupModel extends Disposable {
 				targetIndex = this.editors.length;
 			}
 
+			// Insert in Order
+			else if (this.editorOpenPositioning === EditorOpenPositioning.SORT) {
+				targetIndex = this.editors.length;
+
+				const order = this.editorOpenPositioningSortOrder === 'asc' ? (v: number) => v <= 0 : (v: number) => v > 0;
+				const name = candidate.getTitle(Verbosity.SHORT);
+				const localPath = candidate.getTitle(Verbosity.MEDIUM);
+				const absolutePath = candidate.getTitle(Verbosity.LONG);
+
+				let sort: (i: number) => number = () => -1;
+
+				// Sort by name then, if editor is local, sort by local path; if not, sort by absolute path
+				if (this.editorOpenPositioningSortRule === 'name-local') {
+					sort = (i: number) => {
+						const lc = name.localeCompare(this.editors[i].getTitle(Verbosity.SHORT));
+						if (lc === 0) {
+							if (name === localPath) {
+								if (order(localPath.localeCompare(this.editors[i].getTitle(Verbosity.MEDIUM)))) {
+									return i;
+								}
+							} else if (order(absolutePath.localeCompare(this.editors[i].getTitle(Verbosity.LONG)))) {
+								return i;
+							}
+
+							return -1;
+						} else {
+							return order(lc) ? i : -1;
+						}
+					};
+				}
+
+				// Sort by name then, by absolute path
+				else if (this.editorOpenPositioningSortRule === 'name-absolute') {
+					sort = (i: number) => {
+						const lc = name.localeCompare(this.editors[i].getTitle(Verbosity.SHORT));
+						if (lc === 0) {
+							return order(absolutePath.localeCompare(this.editors[i].getTitle(Verbosity.LONG))) ? i : -1;
+						} else {
+							return order(lc) ? i : -1;
+						}
+					};
+				}
+
+				// Sort by absolute path
+				else if (this.editorOpenPositioningSortRule === 'absolute') {
+					sort = (i: number) => order(absolutePath.localeCompare(this.editors[i].getTitle(Verbosity.LONG))) ? i : -1;
+				}
+
+				let t;
+				for (let i = 0; i < this.editors.length; i++) {
+					if ((t = sort(i)) !== -1) {
+						targetIndex = t;
+						break;
+					}
+				}
+			}
+
 			// Insert to LEFT or RIGHT of active editor
 			else {
 
