diff --git a/build/gulpfile.reh.js b/build/gulpfile.reh.js
index e0df76f..a0d4889 100644
--- a/build/gulpfile.reh.js
+++ b/build/gulpfile.reh.js
@@ -321,3 +321,3 @@ function packageTask(type, platform, arch, sourceFolderName, destinationFolderNa
 		const productJsonStream = gulp.src(['product.json'], { base: '.' })
-			.pipe(json({ commit, date: readISODate('out-build'), version, release }))
+			.pipe(json({ commit, date: readISODate('out-build'), version, release, serverDownloadUrlTemplate: 'https://github.com/zokugun/MrCode/releases/download/${version}.${release}/MrCode-reh-${os}-${arch}-${version}.${release}.tar.gz' }))
 			.pipe(es.through(function (file) {
diff --git a/build/gulpfile.vscode.js b/build/gulpfile.vscode.js
index a63f693..c0eb236 100644
--- a/build/gulpfile.vscode.js
+++ b/build/gulpfile.vscode.js
@@ -285,3 +285,3 @@ function packageTask(platform, arch, sourceFolderName, destinationFolderName, op
 		const productJsonStream = gulp.src(['product.json'], { base: '.' })
-			.pipe(json({ commit, date: readISODate('out-build'), checksums, version, release }))
+			.pipe(json({ commit, date: readISODate('out-build'), checksums, version, release, serverDownloadUrlTemplate: 'https://github.com/zokugun/MrCode/releases/download/${version}.${release}/MrCode-reh-${os}-${arch}-${version}.${release}.tar.gz' }))
 			.pipe(es.through(function (file) {
